src/codeu/chat/client/commandline/Chat.java

```diff
@@ -14,310 +14,384 @@
+  // stack. When a command wants to go to another panel, it will add a new
+  // panel to the top of the stack. When a command wants to go to the previous
+  // panel all it needs to do is pop the top panel.
+  private final Stack<Panel> panels = new Stack<>();
+
+  public Chat(Context context) {
+    this.panels.push(createRootPanel(context));
   }
 
-  // Print help message.
-  private static void help() {
-    System.out.println("Chat commands:");
-    System.out.println("   exit      - exit the program.");
-    System.out.println("   help      - this help message.");
-    System.out.println("   sign-in <username>  - sign in as user <username>.");
-    System.out.println("   sign-out  - sign out current user.");
-    System.out.println("   current   - show current user, conversation, message.");
-    System.out.println("User commands:");
-    System.out.println("   u-add <name>  - add a new user.");
-    System.out.println("   u-list-all    - list all users known to system.");
-    System.out.println("Conversation commands:");
-    System.out.println("   c-add <title>    - add a new conversation.");
-    System.out.println("   c-list-all       - list all conversations known to system.");
-    System.out.println("   c-select <index> - select conversation from list.");
-    System.out.println("Message commands:");
-    System.out.println("   m-add <body>     - add a new message to the current conversation.");
-    System.out.println("   m-list-all       - list all messages in the current conversation.");
-    System.out.println("   m-next <index>   - index of next message to view.");
-    System.out.println("   m-show <count>   - show next <count> messages.");
-  }
+  // HANDLE COMMAND
+  //
+  // Take a single line of input and parse a command from it. If the system
+  // is willing to take another command, the function will return true. If
+  // the system wants to exit, the function will return false.
+  //
+  public boolean handleCommand(String line) {
 
-  // Prompt for new command.
-  private void promptForCommand() {
-    System.out.print(PROMPT);
-  }
-
-  // Parse and execute a single command.
-  private void doOneCommand(Scanner lineScanner) {
+    final Scanner tokens = new Scanner(line.trim());
 
-    final Scanner tokenScanner = new Scanner(lineScanner.nextLine());
-    if (!tokenScanner.hasNext()) {
-      return;
-    }
-    final String token = tokenScanner.next();
+    final String command = tokens.hasNext() ? tokens.next() : "";
 
-    if (token.equals("exit")) {
+    // Because "exit" and "back" are appliable to every panel, handle
+    // those commands here to avoid having to implement them for each
+    // panel.
 
-      alive = false;
+    if ("exit".equals(command)) {
+      // The user does not want to process anymore commands
+      return false;
+    }
 
-    } else if (token.equals("help")) {
+    // Do not allow the root panel to be removed so make sure there
+    // are enough panels before popping.
```

Comment:
Everything after "Do not allow the root panel to be removed" is kind of obvious.